### PR TITLE
Implement S3871: Exceptions should be public - for VB.NET 

### LINKED ISSUE #6279: Implement S3871: Exceptions should be public - for VB.NET
Implement S3871- Exceptions should be public - for VB.NET

### PR DESCRIPTION
**Pull request by Corniel**
Fixes #6279

Moved logic to a base class, and added VB.NET tests.

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
@Corniel can you rebase please?

**Comment by Corniel** [comment]
Rebase done.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Looks quite good, will you create the rspec PR too?

File: analyzers/src/SonarAnalyzer.Common/Rules/ExceptionsShouldBePublicBase.cs
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/ExceptionsShouldBePublicBase.cs:40]: @pavel-mikula-sonarsource I'm not sure. I didn't want to change the behavior (yet). There are testcases that allow to have non-public exceptions on exceptions derived from others.

My main goal was to get this one working for both lanuages first.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ExceptionsShouldBePublicTest.cs
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/ExceptionsShouldBePublicTest.cs:39]: I can not recall clearly; I thought it had something to do with adding the extra namespace scope.. It's some months ago.  ;)

However, It now works like a charm, so I removed the line.

File: analyzers/src/SonarAnalyzer.Common/Rules/ExceptionsShouldBePublicBase.cs
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/ExceptionsShouldBePublicBase.cs:42]: Your example will never return an `INamedTypeSymbol`, so that is not needed. I added the scenario however.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionsShouldBePublic.cs
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionsShouldBePublic.cs:8]: About the exceptions that are taken into account. The [RSPEC](https://rules.sonarsource.com/csharp/RSPEC-3871) explicitly mentions it:
> This rule ignores Exception types that are not derived directly from `System.Exception`, `System.SystemException`, or `System.ApplicationException`.

It is not explained **why**, but by changing the behavior, this part of the documentation has to be adjusted too.

I'm not sure how many new issues there will be reported amongst customers, and if that is desirable or not, that your call.

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionsShouldBePublic.cs:8]: Ok, let's just keep the UT as is and remove the "FN" comment

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionsShouldBePublic.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExceptionsShouldBePublic.cs:82]:
```suggestion
internal class LastException : MiddleException { }
```

File: analyzers/src/SonarAnalyzer.Common/Rules/ExceptionsShouldBePublicBase.cs
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/ExceptionsShouldBePublicBase.cs:49]: Indeed, we're fine with `.Value.` then

**Review by pavel-mikula-sonarsource — COMMENTED** [review]
Looks good so far, I need to update RSPEC repository and push the rspec scaffolding before running UTs

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

File: analyzers/rspec/vbnet/S3871_vb.net.html
Comment (Corniel) [inline analyzers/rspec/vbnet/S3871_vb.net.html:1]: I would ague that this should be in the generic part, and that only the compliant and non-compliant should be language specific?

**Review by Corniel — COMMENTED** [review]
LGTM!

**Comment by pavel-mikula-sonarsource** [comment]
Waiting for RSPEC PR to be reviewed and merged: https://github.com/SonarSource/rspec/pull/1375

**Pull request merged**
